### ISSUE TITLE
assertk application in pure JVM module

### ISSUE THREAD

**Issue by realdadfish** [willowtreeapps/assertk#242]
If I apply assertk-jvm-0.19 to a pure JVM module, Gradle 5.6 reports this:

```
   > Could not resolve com.willowtreeapps.assertk:assertk-jvm:0.19.
     Required by:
         project :core-test
      > Cannot choose between the following variants of com.willowtreeapps.assertk:assertk-jvm:0.19:
          - jvm-api
          - jvm-runtime
          - metadata-api
        All of them match the consumer attributes:
          - Variant 'jvm-api' capability com.willowtreeapps.assertk:assertk-jvm:0.19:
              - Unmatched attributes:
                  - Found org.gradle.libraryelements 'jar' but wasn't required.
                  - Found org.gradle.status 'release' but wasn't required.
                  - Found org.gradle.usage 'java-api' but wasn't required.
                  - Found org.jetbrains.kotlin.platform.type 'jvm' but wasn't required.
          - Variant 'jvm-runtime' capability com.willowtreeapps.assertk:assertk-jvm:0.19:
              - Unmatched attributes:
                  - Found org.gradle.libraryelements 'jar' but wasn't required.
                  - Found org.gradle.status 'release' but wasn't required.
                  - Found org.gradle.usage 'java-runtime' but wasn't required.
                  - Found org.jetbrains.kotlin.platform.type 'jvm' but wasn't required.
          - Variant 'metadata-api' capability com.willowtreeapps.assertk:assertk-jvm:0.19:
              - Unmatched attributes:
                  - Found org.gradle.status 'release' but wasn't required.
                  - Found org.gradle.usage 'kotlin-api' but wasn't required.
                  - Found org.jetbrains.kotlin.platform.type 'common' but wasn't required.
```

Should I apply the `jvm-runtime` artifact in this case?

**Comment by realdadfish** [comment]
Hrm... version 0.16 onwards seems to be the culprit, up until 0.15 it works.

**Comment by realdadfish** [comment]
No, not at hand. I'll try to track the issue down myself first, let's see if I find something.

**Comment by realdadfish** [comment]
Ok, I could fix the issue by adding the following snippet in all of my test modules that directly or indirectly depend on assertk:

```groovy
afterEvaluate {
    configurations.findByName('_classStructurekaptKotlin').attributes {
        attribute(Usage.USAGE_ATTRIBUTE, project.objects.named(Usage, Usage.JAVA_API))
    }
}
```

While the documentation at https://docs.gradle.org/current/userguide/dependency_management_attribute_based_matching.html explains in great detail how a library author can specify certain attributes that should eventually be matched by a consumer, it does not contain any hints how to "configure" a certain attribute, hence exposed variant. I'm new to this attribute-based matching; if you have any hints what I could do better instead on the client side to request the right variant (`jvm-runtime`) then I'd be very grateful :)

**Comment by evant** [comment]
You really shouldn't have to configure anything, that's why I was asking for a sample project to try to reproduce your issue.

**Comment by realdadfish** [comment]
I've came up with a minimal example project - the culprit seems to be kapt, if this plugin is not applied, Gradle won't spit out the error: https://github.com/realdadfish/assertk-issue242

**Comment by evant** [comment]
Hm, I was able to reproduce this with a trivial mulitplatform module:

```groovy
plugins {
    id 'org.jetbrains.kotlin.multiplatform'
}

repositories {
    mavenCentral()
    jcenter()
}

kotlin {
    jvm()
}
```

**Comment by evant** [comment]
Looks like a bug in kapt, will investigate if there's a way to work around it https://youtrack.jetbrains.com/issue/KT-31641

**Comment by evant** [comment]
You can work around it with
```groovy
configurations.all { configuration ->
    // Workaround for kapt bug with MPP dependencies
    // https://youtrack.jetbrains.com/issue/KT-31641
    if (name.contains('kapt')) {
        attributes.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.class, Usage.JAVA_RUNTIME))
    }
}
```
adding this info to the README

**Comment by evant** [comment]
I think this might be fixed in 1.3.60 if anyone wants to test.

**Comment by evant** [comment]
Verified fixed in 1.3.60
